### PR TITLE
Added tags support

### LINKED ISSUE #354: new feature: tags
We are introducing tags in alchemy - basically a many-to-many relation between proposals and tag names.

We will implement this by:

- saving tags as a list on the ipfs "description", which will look like:
` ` `
{
title: "some tittle" ,
description: " ...." ,
tags: ["tag1" , " tag2" , "tag2" ]
}` 
- for the UI, we need to be able to: (a) list the tags with each proposal (b) get the complete list of tags and (c) search for proposals by tag

These data need to be indexed by the subgraph The most straightfoward solution seems to be:

- have a new `Tag` entity with attributes `id`  and `name` (or we can even use the ` name` as the id)
- add a property `tags`  to the Proposal entity, which refres to a list of Tag instances

### PR DESCRIPTION
**Pull request by ben-kaufman**
Fix #354

### CODE REVIEW

**Comment by jellegerbrandy** [comment]
This looks good but is only half of the spec from issue #354. 

`have a new Tag entity with attributes id and name (or we can even use the name as the id)` 

We need a way to get a list of all the tags, probably by

**Comment by jellegerbrandy** [comment]
Sorry about that, I somehow overlooked half of the code (?), for me this is good to go (although it would be even better with a test that queryes the `tags`  collection :-)

File: src/domain/proposal.ts [hunk 1]
```diff
@@ -1,6 +1,6 @@
 import { Address, BigDecimal, BigInt, Bytes, crypto, ipfs, json, JSONValueKind, store } from '@graphprotocol/graph-ts';
 import { GenesisProtocol } from '../types/GenesisProtocol/GenesisProtocol';
-import { ControllerScheme, DAO, GenesisProtocolParam, Proposal } from '../types/schema';
+import { ControllerScheme, DAO, GenesisProtocolParam, Proposal, Tag } from '../types/schema';
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: where `Tag` is being used ?

File: src/domain/proposal.ts [hunk 2]
```diff
@@ -65,6 +65,16 @@ export function getProposalIPFSData(proposal: Proposal): Proposal {
         if (descJson.toObject().get('url') != null) {
           proposal.url = descJson.toObject().get('url').toString();
         }
+        let tags: string[] = [];
+        if (descJson.toObject().get('tags') != null && descJson.toObject().get('tags').kind === JSONValueKind.ARRAY) {
+          let tagsObjects = descJson.toObject().get('tags').toArray();
+          for (let i = 0; i < tagsObjects.length; i++) {
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: does the length should be capped somehow by max number of tags per proposal ? other wise this code might be fragile .

File: src/domain/proposal.ts [hunk 3]
```diff
@@ -65,6 +65,16 @@ export function getProposalIPFSData(proposal: Proposal): Proposal {
...
+            tags.push(tagsObjects[i].toString());
+            let tagEnt = new Tag(tagsObjects[i].toString());
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: how about check if already exist such a Tag  ? to avoid redundant write to db.

File: src/domain/schema.graphql [hunk 4]
```diff
@@ -74,6 +74,7 @@ type Proposal @entity {
 	title: String
 	description: String
 	url: String
+	tags: [String!]
```
Comment (orenyodfat) [inline src/domain/schema.graphql]: tags: [Tag!] ?

File: src/domain/proposal.ts [hunk 5]
```diff
@@ -65,6 +65,16 @@ export function getProposalIPFSData(proposal: Proposal): Proposal {
         if (descJson.toObject().get('url') != null) {
           proposal.url = descJson.toObject().get('url').toString();
         }
+        let tags: string[] = [];
+        if (descJson.toObject().get('tags') != null && descJson.toObject().get('tags').kind === JSONValueKind.ARRAY) {
+          let tagsObjects = descJson.toObject().get('tags').toArray();
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: looks a bit cleaner  : 
```
let tagsData =  descJson.toObject().get('tags');
if (tagsData != null &&
      tagsData.kind === JSONValueKind.ARRAY) {
let tagsObjects = tagsData.toArray(); 
```

File: src/domain/proposal.ts [hunk 6]
```diff
@@ -65,6 +65,19 @@ export function getProposalIPFSData(proposal: Proposal): Proposal {
...
           proposal.url = descJson.toObject().get('url').toString();
         }
+        let tags: string[] = [];
+        let tagsData =  descJson.toObject().get('tags');
+        if (tagsData != null && tagsData.kind === JSONValueKind.ARRAY) {
+          let tagsObjects = tagsData.toArray();
+          for (let i = 0; i < tagsObjects.length; i++) {
+            tags.push(tagsObjects[i].toString());
+            let tagEnt = new Tag(tagsObjects[i].toString());
+            if (Tag.load(tagsObjects[i].toString()) == null) {
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: this check should move to line 73 ?

Comment (ben-kaufman) [reply src/domain/proposal.ts]: No... we want the tag in the list of the proposal anyway

File: src/domain/proposal.ts (hunk 2 above)
Comment (orenyodfat) [reply src/domain/proposal.ts]: @jellegerbrandy does capping that with 100 tags per proposals sound reasonable ?

Comment (orenyodfat) [reply src/domain/proposal.ts]: @ben-kaufman please cap the number of tags per proposal with the magical number 100

File: src/domain/proposal.ts [hunk 7]
```diff
@@ -70,6 +70,9 @@ export function getProposalIPFSData(proposal: Proposal): Proposal {
         if (tagsData != null && tagsData.kind === JSONValueKind.ARRAY) {
           let tagsObjects = tagsData.toArray();
           for (let i = 0; i < tagsObjects.length; i++) {
+            if (i === 100) {
+              break;
+            }
```
Comment (orenyodfat) [inline src/domain/proposal.ts]: `let tagsObjects = tagsData.toArray();
  let tagsLength = tagsObjects.length < 100 ?  tagsObjects.length : 100; 
 for (let i = 0; i < tagsLength; i++) {
...
`

Comment (orenyodfat) [reply src/domain/proposal.ts]: cleaner ?

**Pull request merged**